### PR TITLE
Compute message fee in USD

### PR DESCRIPTION
**Pull request by rstout**
Adds a real implementation for the MessageFeeUSD18Calculator iface that gets the LINK price using CCIPReader and also does fee boosting.

### CODE REVIEW

File: execute/exectypes/costly_messages.go [hunk 1]
```diff
@@ -185,3 +202,63 @@ func (n *StaticMessageExecCostUSD18Calculator) MessageExecCostUSD18(
...
+	messageTimeStamps map[cciptypes.Bytes32]time.Time,
+) (map[cciptypes.Bytes32]plugintypes.USD18, error) {
+	linkPriceUSD, err := c.ccipReader.LinkPriceUSD(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get LINK price in USD: %w", err)
+	}
+
+	messageFees := make(map[cciptypes.Bytes32]plugintypes.USD18)
+	for _, msg := range messages {
+		feeUSD18 := big.NewInt(0).Mul(linkPriceUSD.Int, msg.FeeValueJuels.Int)
```
Comment (winder) [inline execute/exectypes/costly_messages.go:235]: nit: The big API is a little weird, I thought you were multiply everything by 0. It seems like the authors intended for you to use new, [some examples are in the tests](https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/math/big/int_test.go;l=720).
```suggestion
		feeUSD18 := new(big.Int).Mul(linkPriceUSD.Int, msg.FeeValueJuels.Int)
```

File: execute/exectypes/costly_messages.go [hunk 2]
```diff
@@ -185,3 +202,63 @@ func (n *StaticMessageExecCostUSD18Calculator) MessageExecCostUSD18(
...
+// waitBoostedFee boosts the given fee according to the time passed since the msg was sent.
+// RelativeBoostPerWaitHour is used to normalize the time diff,
+// it makes our loss taking "smooth" and gives us time to react without a hard deadline.
+// At the same time, messages that are slightly underpaid will start going through after waiting for a little bit.
+//
+// wait_boosted_fee(m) = (1 + (now - m.send_time).hours * RELATIVE_BOOST_PER_WAIT_HOUR) * fee(m)
+func waitBoostedFee(waitTime time.Duration, fee *big.Int, relativeBoostPerWaitHour float64) *big.Int {
+	k := 1.0 + waitTime.Hours()*relativeBoostPerWaitHour
+
+	boostedFee := big.NewFloat(0).Mul(big.NewFloat(k), new(big.Float).SetInt(fee))
```
Comment (winder) [inline execute/exectypes/costly_messages.go:260]: Same suggestion as above
```suggestion
	boostedFee := new(big.Float).Mul(big.NewFloat(k), new(big.Float).SetInt(fee))
```

**Review by winder — APPROVED** [review]
Looks good, just had a couple minor comments

File: execute/exectypes/costly_messages.go (hunk 1 above)
Comment (rstout) [reply execute/exectypes/costly_messages.go:235]: Good to know, thanks! I think I copied `big.NewInt(0)` from the v1 Exec plugin.

File: execute/exectypes/costly_messages.go [hunk 3]
```diff
@@ -185,3 +202,63 @@ func (n *StaticMessageExecCostUSD18Calculator) MessageExecCostUSD18(
...
+var _ MessageFeeE18USDCalculator = &CCIPMessageFeeUSD18Calculator{}
+
+// MessageFeeUSD18 Returns a map from message ID to the message's fee in USD18s.
+func (c *CCIPMessageFeeUSD18Calculator) MessageFeeUSD18(
+	ctx context.Context,
+	messages []cciptypes.Message,
+	messageTimeStamps map[cciptypes.Bytes32]time.Time,
+) (map[cciptypes.Bytes32]plugintypes.USD18, error) {
+	linkPriceUSD, err := c.ccipReader.LinkPriceUSD(ctx)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (rstout) [reply execute/exectypes/costly_messages.go:229]: I don't think special handling should be needed

File: execute/exectypes/costly_messages.go [hunk 4]
```diff
@@ -185,3 +202,63 @@ func (n *StaticMessageExecCostUSD18Calculator) MessageExecCostUSD18(
...
+	messageFees := make(map[cciptypes.Bytes32]plugintypes.USD18)
+	for _, msg := range messages {
+		feeUSD18 := big.NewInt(0).Mul(linkPriceUSD.Int, msg.FeeValueJuels.Int)
+		timestamp, ok := messageTimeStamps[msg.Header.MessageID]
+		if !ok {
+			// If a timestamp is missing we can't do fee boosting, but we still record the fee. In the worst case, the
+			// message will not be executed (as it will be considered too costly).
+			c.lggr.Warnw("missing timestamp for message", "messageID", msg.Header.MessageID)
+		} else {
+			feeUSD18 = waitBoostedFee(c.now().Sub(timestamp), feeUSD18, c.relativeBoostPerWaitHour)
```
[the remark this answers is not in the document]
Comment (rstout) [reply execute/exectypes/costly_messages.go:242]: Maybe, but I found this way easier to test.

File: execute/exectypes/costly_messages.go (hunk 2 above)
Comment (rstout) [reply execute/exectypes/costly_messages.go:260]: Curiously, these two are not equivalent :\

**Pull request merged**